src/gfx/convert.cpp

```diff
@@ -0,0 +1,1001 @@
+			png_set_add_alpha(png, 0xFFFF, PNG_FILLER_AFTER);
+
+		// Scale 16bpp back to 8 (we don't need all of that precision anyway)
+		if (bitDepth == 16)
+			png_set_scale_16(png);
+		else if (bitDepth < 8)
+			png_set_packing(png);
+
+		// Set interlace handling (MUST be done before `png_read_update_info`)
+		int nbPasses = png_set_interlace_handling(png);
+
+		// Update `info` with the transformations
+		png_read_update_info(png, info);
+		// These shouldn't have changed
+		assert(png_get_image_width(png, info) == width);
+		assert(png_get_image_height(png, info) == height);
+		// These should have changed, however
+		assert(png_get_color_type(png, info) == PNG_COLOR_TYPE_RGBA);
+		assert(png_get_bit_depth(png, info) == 8);
+
+		// Now that metadata has been read, we can process the image data
+
+		std::vector<png_byte> row(png_get_rowbytes(png, info));
+
+		if (interlaceType == PNG_INTERLACE_NONE) {
+			for (png_uint_32 y = 0; y < height; ++y) {
+				png_read_row(png, row.data(), nullptr);
+
+				for (png_uint_32 x = 0; x < width; ++x) {
+					Rgba rgba(row[x * 4], row[x * 4 + 1], row[x * 4 + 2], row[x * 4 + 3]);
+
+					colors.registerColor(rgba);
+					pixel(x, y) = rgba;
+				}
+			}
+		} else {
+			// For interlace to work properly, we must read the image `nbPasses` times
+			for (int pass = 0; pass < nbPasses; ++pass) {
+				// The interlacing pass must be skipped if its width or height is reported as zero
+				if (PNG_PASS_COLS(width, pass) == 0 || PNG_PASS_ROWS(height, pass) == 0) {
+					continue;
+				}
+
+				png_uint_32 xStep = 1u << PNG_PASS_COL_SHIFT(pass);
+				png_uint_32 yStep = 1u << PNG_PASS_ROW_SHIFT(pass);
+
+				for (png_uint_32 y = PNG_PASS_START_ROW(pass); y < height; y += yStep) {
+					png_bytep ptr = row.data();
+
+					png_read_row(png, ptr, nullptr);
+					for (png_uint_32 x = PNG_PASS_START_COL(pass); x < width; x += xStep) {
+						Rgba rgba(ptr[0], ptr[1], ptr[2], ptr[3]);
+
+						colors.registerColor(rgba);
+						pixel(x, y) = rgba;
+						ptr += 4;
+					}
+				}
+			}
+		}
+
+		// We don't care about chunks after the image data (comments, etc.)
+		png_read_end(png, nullptr);
+	}
+
+	~Png() { png_destroy_read_struct(&png, &info, nullptr); }
+
+	class TilesVisitor {
+		Png const &_png;
+		bool const _columnMajor;
+		uint32_t const _width, _height;
+		uint32_t const _limit = _columnMajor ? _height : _width;
+
+	public:
+		TilesVisitor(Png const &png, bool columnMajor, uint32_t width, uint32_t height)
+		    : _png(png), _columnMajor(columnMajor), _width(width), _height(height) {}
+
+		class Tile {
+			Png const &_png;
+			uint32_t const _x, _y;
+
+		public:
+			Tile(Png const &png, uint32_t x, uint32_t y) : _png(png), _x(x), _y(y) {}
+
+			Rgba pixel(uint32_t xOfs, uint32_t yOfs) const {
+				return _png.pixel(_x + xOfs, _y + yOfs);
+			}
+		};
+
+	private:
+		struct iterator {
+			TilesVisitor const &parent;
+			uint32_t const limit;
+			uint32_t x, y;
+
+		public:
```

Comment:
`struct`s are public by default.
```suggestion
		public:
```

Reply:
A remnant of when I had to make the class `public`, but I managed to backtrack.